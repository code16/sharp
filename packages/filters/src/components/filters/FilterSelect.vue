<template>
    <div class="SharpFilterSelect"
          :class="{
              'SharpFilterSelect--open':opened,
              'SharpFilterSelect--empty':empty,
              'SharpFilterSelect--multiple':multiple,
              'SharpFilterSelect--searchable':searchable
          }"
          tabindex="0"
    >
        <!-- dropdown & search input -->
        <Autocomplete
            class="SharpFilterSelect__select"
            :value="autocompleteValue"
            :local-values="values"
            :search-keys="searchKeys"
            :list-item-template="template"
            :placeholder="l('entity_list.filter.search_placeholder')"
            :multiple="multiple"
            :hide-selected="multiple"
            :allow-empty="!required"
            :preserve-search="false"
            :show-pointer="false"
            :searchable="searchable"
            no-result-item
            mode="local"
            ref="autocomplete"
            @multiselect-input="handleAutocompleteInput"
            @close="close"
        />

        <FilterControl :label="label" no-caret @click="handleClick">
            <!-- value text & tags -->
            <Select
                class="SharpFilterSelect__select"
                :value="value"
                :options="values"
                :multiple="multiple"
                :clearable="!required"
                :inline="false"
                placeholder=" "
                ref="select"
                @input="handleSelect"
            />
        </FilterControl>
    </div>
</template>

<script>
    import { Autocomplete, Select } from 'sharp-form'
    import { Localization } from 'sharp/mixins';
    import FilterControl from '../FilterControl';
    
    export default {
        name: 'SharpFilterSelect',
        mixins: [Localization],
        components: {
            Select,
            Autocomplete,
            FilterControl,
        },
        props: {
            label : {
                type: String,
                required: true
            },
            values: {
                type: Array,
                required: true
            },
            value: {
                type: [String, Number, Array],
            },
            multiple: Boolean,
            required: Boolean,
            searchable: Boolean,
            searchKeys: Array,
            template: String,
        },
        data() {
            return {
                opened: false,
                // dev only, check it in devtools to keep dropdown open
                debug: false,
            }
        },
        computed: {
            optionById() {
                return this.values.reduce((res, v)=> ({
                    ...res, [v.id]: v
                }), {});
            },
            empty() {
                return this.value == null || this.multiple && !this.value.length;
            },
            autocompleteValue() {
                return this.multiple ? (this.value||[]).map(value=>this.optionById[value]) : this.optionById[this.value];
            }
        },
        methods: {
            handleSelect(value) {
                this.$emit('input', value);
            },
            handleAutocompleteInput(value) {
                this.$emit('input', this.multiple ? value.map(v=>v.id) : (value||{}).id);
            },
            handleClick() {
                if(this.opened) {
                    this.close();
                } else {
                    this.open();
                }
            },
            open() {
                this.opened = true;
                this.$emit('open');
                this.$nextTick(this.showDropdown);
            },
            close() {
                if(this.debug) {
                    return;
                }
                this.opened = false;
                this.$emit('close');
                this.$nextTick(this.blur);
            },
            showDropdown() {
                let { autocomplete:{ $refs: { multiselect } } } = this.$refs;
                multiselect.activate();

                if(this.debug) {
                    this.unwatch && this.unwatch();
                    this.unwatch = multiselect.$watch('isOpen', function (isOpen) { if(!isOpen)this.isOpen = true; }, { sync: true });
                }
            },
            blur() {
                let { select:{ $refs: { multiselect } } } = this.$refs;
                multiselect.deactivate();
            },
        }
    }
</script>