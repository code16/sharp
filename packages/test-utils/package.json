{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}