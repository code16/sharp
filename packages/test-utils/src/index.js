export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}