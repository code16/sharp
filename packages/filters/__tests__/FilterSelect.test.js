import FilterSelect from '../src/components/filters/FilterSelect.vue';
import { Select } from 'sharp-form';
import { shallowMount } from '@vue/test-utils';
import { MockI18n } from "@sharp/test-utils";
import Vue from 'vue';


describe('filter-select', ()=>{
    Vue.use(MockI18n, { mockFn:true });

    function createWrapper(options={}) {
        return shallowMount(FilterSelect, {
            ...options,
            propsData : {
                filterKey: 'job',
                label: 'Web job',
                values: [
                    { id:1, label:'front' },
                    { id:2, label:'back' },
                ],
                ...options.propsData
            },
        });
    }

    function findSelect(wrapper) {
        return wrapper.find(Select);
    }

    test('can mount "empty" FilterSelect', ()=>{
        expect(createWrapper().html()).toMatchSnapshot();
    });

    test('can mount "valuated" FilterSelect', ()=>{
        expect(createWrapper({propsData: { value:1 }}).html()).toMatchSnapshot();
    });

    test('can mount "multiple empty" FilterSelect', ()=>{
        expect(createWrapper({ propsData:{multiple:true} }).html()).toMatchSnapshot();
    });

    test('can mount "multiple valuated" FilterSelect', ()=>{
        expect(createWrapper({ propsData:{multiple:true, value: [1,2]} }).html()).toMatchSnapshot();
    });

    test('has select', ()=>{
        expect(findSelect(createWrapper()).isVueInstance()).toBe(true);
    });

    test('expose appropriate props to select', ()=>{
        const wrapper = createWrapper({
            propsData: {
                multiple: true,
                required: true,
                value: [1,2]
            }
        });
        const select = findSelect(wrapper);

        expect(select.vm.$props).toMatchObject({
            value: [1,2],
            options: [
                { id:1, label:'front' },
                { id:2, label:'back' },
            ],
            multiple: true,
            clearable: false,
            inline: false,
        });
    });

    test('call appropriate handlers', async ()=>{
        const wrapper = createWrapper({
            created() {
                this.handleSelect = jest.fn();
            }
        });
        const select = findSelect(wrapper);

        select.vm.$emit('input');
        expect(wrapper.vm.handleSelect).toHaveBeenCalled();
    });

    test('empty', ()=>{
        const wrapper = createWrapper();

        expect(wrapper.vm.empty).toBe(true);
        wrapper.setProps({ value:1 });
        expect(wrapper.vm.empty).toBe(false);
        wrapper.setProps({ value:[], multiple:true });
        expect(wrapper.vm.empty).toBe(true);
        wrapper.setProps({ value:[1], multiple:true });
        expect(wrapper.vm.empty).toBe(false);
    });

    test('emit input', ()=>{
        const wrapper = createWrapper();

        wrapper.vm.handleSelect(1);
        expect(wrapper.emitted('input')).toEqual([[1]]);
    });
});