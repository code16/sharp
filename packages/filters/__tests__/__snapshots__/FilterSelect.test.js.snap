// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter-select can mount "empty" FilterSelect 1`] = `
<div tabindex="0" class="SharpFilterSelect SharpFilterSelect--empty" filterkey="job">
	<autocomplete-stub mode="local" localvalues="[object Object],[object Object]" placeholder="{{ entity_list.filter.search_placeholder }}" remotesearchattribute="query" itemidattribute="id" searchminchars="1" searchkeys="value" noresultitem="true" allowempty="true"
	    class="SharpFilterSelect__select"></autocomplete-stub>
	<filtercontrol-stub label="Web job" nocaret="true">
		<select-stub options="[object Object],[object Object]" display="dropdown" clearable="true" placeholder=" " class="SharpFilterSelect__select"></select-stub>
	</filtercontrol-stub>
</div>
`;

exports[`filter-select can mount "multiple empty" FilterSelect 1`] = `
<div tabindex="0" class="SharpFilterSelect SharpFilterSelect--empty SharpFilterSelect--multiple" filterkey="job">
	<autocomplete-stub value="" mode="local" localvalues="[object Object],[object Object]" placeholder="{{ entity_list.filter.search_placeholder }}" remotesearchattribute="query" itemidattribute="id" searchminchars="1" searchkeys="value" noresultitem="true"
	    multiple="true" hideselected="true" allowempty="true" class="SharpFilterSelect__select"></autocomplete-stub>
	<filtercontrol-stub label="Web job" nocaret="true">
		<select-stub options="[object Object],[object Object]" multiple="true" display="dropdown" clearable="true" placeholder=" " class="SharpFilterSelect__select"></select-stub>
	</filtercontrol-stub>
</div>
`;

exports[`filter-select can mount "multiple valuated" FilterSelect 1`] = `
<div tabindex="0" class="SharpFilterSelect SharpFilterSelect--multiple" filterkey="job">
	<autocomplete-stub value="[object Object],[object Object]" mode="local" localvalues="[object Object],[object Object]" placeholder="{{ entity_list.filter.search_placeholder }}" remotesearchattribute="query" itemidattribute="id" searchminchars="1" searchkeys="value"
	    noresultitem="true" multiple="true" hideselected="true" allowempty="true" class="SharpFilterSelect__select"></autocomplete-stub>
	<filtercontrol-stub label="Web job" nocaret="true">
		<select-stub value="1,2" options="[object Object],[object Object]" multiple="true" display="dropdown" clearable="true" placeholder=" " class="SharpFilterSelect__select"></select-stub>
	</filtercontrol-stub>
</div>
`;

exports[`filter-select can mount "valuated" FilterSelect 1`] = `
<div tabindex="0" class="SharpFilterSelect" filterkey="job">
	<autocomplete-stub value="[object Object]" mode="local" localvalues="[object Object],[object Object]" placeholder="{{ entity_list.filter.search_placeholder }}" remotesearchattribute="query" itemidattribute="id" searchminchars="1" searchkeys="value" noresultitem="true"
	    allowempty="true" class="SharpFilterSelect__select"></autocomplete-stub>
	<filtercontrol-stub label="Web job" nocaret="true">
		<select-stub value="1" options="[object Object],[object Object]" display="dropdown" clearable="true" placeholder=" " class="SharpFilterSelect__select"></select-stub>
	</filtercontrol-stub>
</div>
`;
